### ISSUE TITLE
Future-Proof Nametag Generation

### ISSUE THREAD

**Issue by miterion** [d120/pyophase#339]
The current nametag generation functionality is based around an old script, which does not accommodate to current requirements. Furthermore, last years quick patches (thanks to @bhaettasch) are only a temporary solution to the new maximum sized rooms problem and limited availability of tutors.

Therefore a new solution should be built which meets the following requirements:
1. Use the existing [Room model](https://github.com/d120/pyophase/blob/778bee856c198c5f658add124bb2314315ff350b/ophasebase/models/location.py#L35-L64) or enhance it
2. Allow organizers to enter all tutor group appointments per Ophase
3. Allow organizers to enter the room availability
4. From the data entered in the last two steps, generate a room plan for the tutor groups
5. Use that data to place freshmen into groups according to the room capacity

For step 3. it would be nice to model the ui to closely match the format in which the organizers get the available room list from the department. @ckleemann already agreed to work on this, but input from previous head of ophases would be appreciated here (@bhaettasch , @max-weller , @jlauinger for example)

**Comment by ckleemann** [comment]
I talked to @pollti last night. There a plans to change the procedure how freshmen are assigned to tutor groups. The main idea is to only a assign a freshman to a group if they show up. With this new procedure the number of groups with a few freshman will be minimized. To assign a freshman to a group they want to use stickers etc.

I therefore would postpone this issue until the procedure for the next ophase is defined.

**Comment by pollti** [comment]
We would like to try printing nametags and names separately.
So, our current change requests would be:
* nametags in the old design, but
  * no name (and maybe a red note that the name should be added („Bitte hier Namen aufkleben. Solltest du keinen Namensaufkleber haben, wende dich bitte an den Helpdesk.“).
  * the ordering of nametags should be adjusted, so that all groups have about the same number of members up to the maximal number for single groups (if groups A and B have capacity 4 and C has capacity 2 we want them ordered as ABCABCABAB); maybe ordering can also be done manually later, but the correct number would be great.
* nametags printable in stickers (should hopefully be printable with PBody).
* master nametags should have the name printed in before, but currently these are not generatable by pyOphase, so it would be nice to have this feature in the future.

Do you think these changes are realistic?

Thanks on behalf of
Ophasenleitung WT 2020

**Comment by ckleemann** [comment]
Yes it sounds realistic. Regarding the ordering I think this can be done in software but since there are many steps which require manual labor I think it is better to provide some assistance. Since the goal of this effort is to have a maximum number of groups which are filled with as many persons as possible I would recommend a different ordering:

I would order them in a hot rotation way: So if there are e.g. 20 groups (A - T ) I would order them from small to big and would only shuffle 5 at the same time. So if A – C has 4 slots, E – F 6 slots and G – T has 20 slots I would order them: ABCEF 4 times and then replace ABC by GEH and so on. By the problem of friends appearing at the same time is solved, the small groups are filled to the maximum and big groups probably are less filled. By writing this I also think about to use a flexible shuffle number so you don’t get group 19 filled to the max and 20 is never used. I will think a little about it. Do you have any percentage how many students did not show up the last time?

**Comment by pollti** [comment]
To be honest, I currently don't recognice the advantage of this ordering.

I think our goals should be
* as close to equally sized groups as possible
* different groups for people showing up together
* not all people coming early/late in the same groups

If i don't miss a point here, (AB…T)⁴(EF…T)²(GH…T)¹⁴ would fill up all groups equally (up to their maximum size).
I would expect about 50% to show up.
